sdk/src/main/scala/com/horizen/validation/MainchainBlockReferenceValidator.scala

```diff
@@ -0,0 +1,96 @@
+  // McData:      10      -       11        12,13     14
+  private def validateBlock(verifiedBlock: SidechainBlock, history: SidechainHistory): Unit = {
+    if(verifiedBlock.mainchainBlockReferencesData.isEmpty)
+      return
+
+    // Loop back in history and found MainchainHeaders that wait for MainchainBlockReferenceData to be included.
+    var missedMainchainReferenceDataHeaderHashesInfo: Seq[(MainchainHeaderHash, ModifierId)] = Seq()
+
+    breakable {
+      var blockId: ModifierId = verifiedBlock.parentId
+      var bestMainchainReferenceDataHeaderHashOpt: Option[MainchainHeaderHash] = None
+      while (true) {
+        val blockInfo: SidechainBlockInfo = history.blockInfoById(blockId)
+        // Prepend the list with MainchainHeaders info (hash and containing SidechainBlock id)
+        missedMainchainReferenceDataHeaderHashesInfo = blockInfo.mainchainHeaderHashes.map((_, blockId)) ++ missedMainchainReferenceDataHeaderHashesInfo
+
+        // Try to get the last MainchainReferenceData HeaderHash included into the chain
+        if(bestMainchainReferenceDataHeaderHashOpt.isEmpty)
+          bestMainchainReferenceDataHeaderHashOpt = blockInfo.mainchainReferenceDataHeaderHashes.lastOption
+
+        if(bestMainchainReferenceDataHeaderHashOpt.isDefined) {
+          // Go back in history until found MainchainHeader corresponding to the best known MainchainReferenceData
+          val index: Int = missedMainchainReferenceDataHeaderHashesInfo.indexWhere(info => info._1 == bestMainchainReferenceDataHeaderHashOpt.get)
+          if (index != -1) {
+            // Drop the Headers info till best known MainchainReferenceData HeaderHash
+            missedMainchainReferenceDataHeaderHashesInfo = missedMainchainReferenceDataHeaderHashesInfo.drop(index + 1)
+            break
+          }
+        }
+        blockId = blockInfo.parentId
+      }
+    }
+
+    // Update header hashes info with verified block MainchainHeaders info
+    missedMainchainReferenceDataHeaderHashesInfo =
+      missedMainchainReferenceDataHeaderHashesInfo ++ verifiedBlock.mainchainHeaders.map(header => (byteArrayToMainchainHeaderHash(header.hash), verifiedBlock.id))
+
+    // Block MainchainBlockReferencesData number must be not greater, then missed ones till verified block
+    if(verifiedBlock.mainchainBlockReferencesData.size > missedMainchainReferenceDataHeaderHashesInfo.size)
+      throw new InvalidMainchainDataException("Block contains more MainchainBlockReferenceData, than expected in the chain.")
+
+    // Collect MainchainHeaders with corresponding MainchainReferenceData into MainchainReferences and verify them.
+    verifiedBlock.mainchainBlockReferencesData.zip(missedMainchainReferenceDataHeaderHashesInfo).foldLeft(verifiedBlock){
+      case (lastRetrievedBlock, (referenceData, (mainchainHeaderHash, containingBlockId))) =>
+        val blockWithMainchainHeader: SidechainBlock = containingBlockId match {
```

Comment:
The whole algorithm rely on having two ordered and consistent lists of mc block hashes for data and headers , then for clarity (even if MainchainBlockReference.semanticValidity already do this)  would be helpful to verify that referenceData.headerHash == mainchainHeaderHash